improver/utilities/field_texture.py

```diff
@@ -0,0 +1,208 @@
+            2. Calculate the number of actual transitions within the neighbourhood,
+               that is the number of cells of value 0 that orthogonally abut cells
+               of value 1.
+            3. Calculate the ratio of actual to potential transitions.
+
+        Ratios approaching 1 indicate that there are many transitions, so the field
+        is highly textured. Ratios close to 0 indicate a smoother field.
+
+        A neighbourhood full of cells of value 1 will return ratios of 0; the
+        diagnostic that has been thresholded to produce the binary field is found
+        everywhere within that neighbourhood, giving a smooth field. At the other
+        extreme, in neighbourhoods in which there are no cells of value 1 the ratio
+        is set to 1.
+
+        Args:
+            cube (iris.cube.Cube):
+                Input data in cube format containing a two-dimensional field
+                of binary data.
+
+            radius (float):
+                Radius for neighbourhood in metres.
+
+        Returns:
+            iris.cube.Cube:
+                A ratio between 0 and 1 of actual transitions over potential transitions.
+        """
+        # Calculate the potential transitions within neighbourhoods
+        potential_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            cube, radius=radius
+        )
+        potential_transitions.data = np.where(
+            cube.data > 0, potential_transitions.data * 4, 0
+        )
+
+        # Calculate the actual transitions for each grid cell of value 1 and
+        # store them in a cube.
+        actual_transitions = potential_transitions.copy(
+            data=self._calculate_transitions(cube.data)
+        )
+
+        # Sum the number of actual transitions within the neighbourhood.
+        actual_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            actual_transitions, radius=radius
+        )
+        actual_transitions.data = np.where(cube.data > 0, actual_transitions.data, 0)
+
+        # Calulate the ratio of actual to potential transitions
+        ratio = np.ones_like(actual_transitions.data)
+        np.divide(
+            actual_transitions.data,
+            potential_transitions.data,
+            out=ratio,
+            where=(cube.data > 0),
+        )
+
+        # Create a new cube to contain the resulting ratio data
+        ratio = create_new_diagnostic_cube(
+            "cell_edge_differences", 1, cube, MANDATORY_ATTRIBUTE_DEFAULTS, data=ratio,
+        )
+        return ratio
+
+    @staticmethod
+    def _calculate_transitions(data):
+        """
+        Identifies actual transitions present in a binary field. These transitions
+        are defined as the number of cells of value zero that directly neighbour
+        cells of value 1. The number of transitions is calculated for all cells
+        of value 1. The number of transitions for cells of value 0 is set to 0.
```

Comment:
Maybe state outright that this is avoid double-counting transitions?

Reply:
Done.